File: library/src/main/java/me/yamlee/jsbridge/widget/dialog/internel/ListDialog.java
package me.yamlee.jsbridge.widget.dialog.internel;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.support.v7.app.AlertDialog;
import android.widget.BaseAdapter;

import java.util.List;

/**
 * 通用列表显示弹框
 *
 * @author yamlee
 */
public class ListDialog {

    public static ListDialog.Builder builder() {
        return new Builder();
    }


    public static class Builder {
        private String title = "";
        private String[] itemArray;
        private List<String> itemList;
        private BaseAdapter mAdapter;
        private int mMaxHeight;
        private int mMaxWidth;

        public ListDialog.Builder setTitle(String title) {
            this.title = title;
            return this;
        }

        /**
         * 设置字符数组的数据源
         *
         * @param itemArray
         */
        public ListDialog.Builder setData(String[] itemArray) {
            this.itemArray = itemArray;
            return this;
        }

        /**
         * 设置字符集合类型的数据源,如果同时设置了两种类型的数据源,优先采用字符数组
         *
         * @param itemList
         */
        public ListDialog.Builder setData(List<String> itemList) {
            this.itemList = itemList;
            return this;
        }

        public ListDialog.Builder setAdapter(BaseAdapter mAdapter) {
            this.mAdapter = mAdapter;
            return this;
        }

        public ListDialog.Builder setMaxHeight(int mMaxHeight) {
            this.mMaxHeight = mMaxHeight;
            return this;
        }

        public ListDialog.Builder setMaxWidth(int mMaxWidth) {
            this.mMaxWidth = mMaxWidth;
            return this;
        }

        public Dialog build(Context context, final PositionClickListener<DialogInterface> listener) {
            String[] data = new String[0];
            if (itemArray != null) {
                data = itemArray;
            } else if (itemList != null) {
                data = itemList.toArray(new String[itemList.size()]);
            } else if (mAdapter != null) {
                Dialog dialog = new AlertDialog.Builder(context).setTitle(title)
                        .setAdapter(mAdapter, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                listener.onClick(which, dialog);
                            }
                        })
                        .create();
                return dialog;
            } else {
                //noinspection HardCodedStringLiteral
                throw new RuntimeException("you should set data for ListDialog");
            }
            Dialog dialog = new AlertDialog.Builder(context).setTitle(title)
                    .setItems(data, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            listener.onClick(which, dialog);
                        }
                    })
                    .create();
            return dialog;
        }

    }


    public interface PositionClickListener<T> {
        void onClick(int position, T t);
    }
}

File: android-jsBridgeAll/src/main/java/me/yamlee/bridge/ui/dialog/internel/ListDialog.kt
package me.yamlee.bridge.ui.dialog.internel

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.support.v7.app.AlertDialog
import android.widget.BaseAdapter

/**
 * 通用列表显示弹框
 *
 * @author yamlee
 */
object ListDialog {

    fun builder(): ListDialog.Builder {
        return Builder()
    }


    class Builder {
        private var title = ""
        private var itemArray: Array<String>? = null
        private var itemList: List<String>? = null
        private var mAdapter: BaseAdapter? = null
        private var mMaxHeight: Int = 0
        private var mMaxWidth: Int = 0

        fun setTitle(title: String): ListDialog.Builder {
            this.title = title
            return this
        }

        /**
         * 设置字符数组的数据源
         *
         * @param itemArray
         */
        fun setData(itemArray: Array<String>): ListDialog.Builder {
            this.itemArray = itemArray
            return this
        }

        /**
         * 设置字符集合类型的数据源,如果同时设置了两种类型的数据源,优先采用字符数组
         *
         * @param itemList
         */
        fun setData(itemList: List<String>): ListDialog.Builder {
            this.itemList = itemList
            return this
        }

        fun setAdapter(mAdapter: BaseAdapter): ListDialog.Builder {
            this.mAdapter = mAdapter
            return this
        }

        fun setMaxHeight(mMaxHeight: Int): ListDialog.Builder {
            this.mMaxHeight = mMaxHeight
            return this
        }

        fun setMaxWidth(mMaxWidth: Int): ListDialog.Builder {
            this.mMaxWidth = mMaxWidth
            return this
        }

        fun build(context: Context, listener: PositionClickListener<DialogInterface>): Dialog {
            var data: Array<String>?
            if (itemArray != null) {
                data = itemArray
            } else if (itemList != null) {
                data = itemList!!.toTypedArray()
            } else if (mAdapter != null) {
                val dialog = AlertDialog.Builder(context).setTitle(title)
                        .setAdapter(mAdapter) { dialog, which -> listener.onClick(which, dialog) }
                        .create()
                return dialog
            } else {

                throw RuntimeException("you should set data for ListDialog")
            }
            val dialog = AlertDialog.Builder(context).setTitle(title)
                    .setItems(data) { dialog, which -> listener.onClick(which, dialog) }
                    .create()
            return dialog
        }

    }


    interface PositionClickListener<T> {
        fun onClick(position: Int, t: T)
    }
}
